### PR TITLE
feat(payment-card): add component Payment Card

### PR DESCRIPTION
**Pull request by tajo**
payment-card

Originally, I thought it could be a sub-export of `baseui/input` but then I saw that `phone-input` is also a standalone component so following the same convention. It's also good for code-splitting (no need to rely on unreliable tree-shaking).

Also waiting for the final sign-off from @nguyensomniac that we can open-source the card icons but should be fine.

### CODE REVIEW

File: src/payment-card/payment-card.js [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+  const number = input.replace(/[^0-9]/gi, '');
+  const validatedValue = valid.number(number);
+  if (validatedValue.card && Array.isArray(validatedValue.card.lengths)) {
+    return number.slice(
+      0,
+      validatedValue.card.lengths[validatedValue.card.lengths.length - 1],
+    );
+  }
+  // CC number NEVER can have more than 19 digits
+  return number.slice(0, 19);
```
Comment (gergelyke) [inline src/payment-card/payment-card.js]: can we limit the length based on the credit card type?

Comment (tajo) [reply src/payment-card/payment-card.js]: Yep, that's what happens right above! This is a fallback when the type is not detected (yet).

Comment (gergelyke) [reply src/payment-card/payment-card.js]: oh, I've overlooked it, great!

File: documentation-site/static/examples/payment-card/controlled.js [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+          onChange={event => this.setState({value: event.target.value})}
+          placeholder="Enter payment card number"
+          value={this.state.value}
+          size={SIZE.default}
+        />
+        <Block as="br" />
+        <PaymentCard
+          onChange={event => this.setState({value: event.target.value})}
+          placeholder="Enter payment card number"
+          value={this.state.value}
```
Comment (nadiia) [inline documentation-site/static/examples/payment-card/controlled.js]: All 3 inputs share the value. Ok to showcase the sizing but may be confusing in the context of functionality.

File: src/payment-card/__tests__/payment-card.test.js [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+  expect(sanitizeNumber(' 4111/ ')).toBe('4111');
+  expect(sanitizeNumber('kkk4-1-1-1abc')).toBe('4111');
+
+  expect(sanitizeNumber('5555555555554444 1234')).toBe('5555555555554444');
+  expect(sanitizeNumber('abc4111 1111 1111 1111 333 444')).toBe(
+    '4111111111111111333',
+  );
+});
+
+//asdasd 5555 5555 5555 4444 1111
```
Comment (nadiia) [inline src/payment-card/__tests__/payment-card.test.js]: nit: remove commented out example

File: src/payment-card/types.js [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+  ENHANCER_POSITION,
+} from '../input/constants.js';
+
+export type AdjoinedT = $Keys<typeof ADJOINED>;
+
+export type SizeT = $Keys<typeof SIZE>;
+
+export type StateTypeT = $Keys<typeof STATE_CHANGE_TYPE>;
+
+export type InternalStateT = {
```
Comment (nadiia) [inline src/payment-card/types.js]: This seems to be not used.

File: src/payment-card/types.js [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+import * as React from 'react';
+import type {OverrideT} from '../helpers/overrides.js';
+import {
+  STATE_CHANGE_TYPE,
+  ADJOINED,
+  SIZE,
+  ENHANCER_POSITION,
+} from '../input/constants.js';
+
+export type AdjoinedT = $Keys<typeof ADJOINED>;
```
Comment (nadiia) [inline src/payment-card/types.js]: This seems to be not used.

**Comment by tajo** [comment]
> Looks good! A question about type exports: do we need all of those that are same as in Input to be recreates and exported in PaymentCard component?

I cleaned up the `types` file. It now reuses all input types and only extends it with the `IconWrapper` override. That's actually the only different between `input` and `payment-card` APIs.

File: src/payment-card/index.js [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+LICENSE file in the root directory of this source tree.
+*/
+// @flow
+export {default as StatefulPaymentCard} from './stateful-payment-card.js';
+export {default as StatefulContainer} from '../input/stateful-container.js';
+export {default as PaymentCard} from './payment-card.js';
+export {default as valid} from 'card-validator';
+export {SIZE} from '../input/index.js';
+// Styled elements
+export {IconWrapper as StyledIconWrapper} from './styled-components.js';
```
Comment (nadiia) [inline src/payment-card/index.js]: Types export is missing.

File: src/payment-card/types.js [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+
+import type {
+  InputComponentsT,
+  InputPropsT,
+  StInputPropsDiffT,
+} from '../input/types.js';
+import type {OverrideT} from '../helpers/overrides.js';
+
+export type PaymentCardComponentsT = {
+  ...InputComponentsT,
```
Comment (chasestarr) [inline src/payment-card/types.js]: fyi spreads from inexact types will mark the 'spread-ed' fields as optional [try-flow](https://flow.org/try/#0C4TwDgpgBAogHgQwMbCgXigbwD5QGYD2BAXFAM7ABOAlgHYDmU2AvgNwBQ7okUAkrREQp0WfEVIUaDKG07doAZTCUICACbxkqDJigA6A5pQAaKACMElUrQCuAWzMRKMjvKhKV6-oK0jdBvW8hYFMLKyhbBycXTiQCWgooHxRSD1UNYL8w0gBGAAZZOITUOmTgVOV0oN8dbKh8tiA)

File: src/payment-card/payment-card.js [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+        CardTypeToComponent[type || 'generic'] || GenericIcon,
+        {
+          size: iconSize[size],
+        },
+      )}
+    </IconWrapper>
+  );
+
+  return (
+    <Input
```
Comment (nudge) [inline src/payment-card/payment-card.js]: some extra attributes may be useful here:
`inputMode: 'numeric'`
`autoComplete: 'cc-number'`

`inputMode` hints to the browser to reveal a numeric keypad - useful for mobile clients
`autoComplete` hints to the browser this field supports card autofill

File: src/payment-card/payment-card.js [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+      size={size}
+      data-baseweb="payment-card"
+      overrides={{
+        ...overrides,
+        Before: BeforeComponent,
+      }}
+      onChange={e => {
+        e.target.value = sanitizeNumber(e.target.value);
```
Comment (nudge) [inline src/payment-card/payment-card.js]: This will probably cause the cursor to jump to the end of the text input if the user tries to edit characters in the middle of the field.

File: src/payment-card/types.js (hunk 7 above)
Comment (tajo) [reply src/payment-card/types.js]: Interesting. Should be fine in this case, since all those props were already optional.

File: src/payment-card/payment-card.js (hunk 8 above)
Comment (tajo) [reply src/payment-card/payment-card.js]: `autoComplete: 'cc-number'` was there already but I've added the `inputMode` as well

File: src/payment-card/payment-card.js (hunk 9 above)
Comment (tajo) [reply src/payment-card/payment-card.js]: fixed, that was a fun one :)

**Comment by chasestarr** [comment]
Could we get some e2e tests to validate the edge-cases we talked about the other day? Unit tests are great, but may not be representative of real usage

ex:
```
type 411122
validate expected number representation
press arrow-left twice
type 33
validate expected number representation 
```

**Comment by tajo** [comment]
> Could we get some e2e tests to validate the edge-cases we talked about the other day? Unit tests are great, but may not be representative of real usage

Yep, added a suite of e2e tests!
